control-plane/p4RuntimeSerializer.cpp

```diff
@@ -639,6 +640,78 @@ getMatchType(cstring matchTypeName) {
     }
 }
 
+/*
+ * This function supports a deeply nested mix of Type_Newtype and typedef.
+ *The function also supported Type_Bits.
+*/
+static uint32_t
+getTypeWidth(const IR::Type* type, TypeMap* typeMap, ReferenceMap* refMap) {
+    uint32_t w = 0;
+    if (type == nullptr)
+        return w;
+
+    LOG3(">> " << type->getP4Type());
+    const IR::Type* newType = type;
+    while (newType->is<IR::Type_Newtype>())
+        newType = newType->to<IR::Type_Newtype>()->type;
+    while (newType->is<IR::Type_Typedef>())
+        newType = newType->to<IR::Type_Typedef>()->type;
+
+    LOG3("After while " << newType->getP4Type());
+    if (newType->is<IR::Type_Bits>()) {
+        w = newType->width_bits();
+        LOG3("First bits " << type->getP4Type());
+        return w;
+    } else if (newType->is<IR::Type_Name>()) {
+        auto n = newType->to<IR::Type_Name>();
+        auto canon = typeMap->getTypeType(n, true);
+        if (canon->is<IR::Type_Bits>()) {
+            auto k = canon->to<IR::Type_Bits>();
+            w = k->width_bits();
+            LOG3("Second bits " << type->getP4Type());
+            return w;
+        } else {
+            auto decl = refMap->getDeclaration(n->path, true);
+            LOG3("Recursion " << n->path);
+            if (decl->is<IR::Type_Newtype>())
+                w = getTypeWidth(decl->to<IR::Type_Newtype>(), typeMap, refMap);
+            else if (decl->is<IR::Type_Typedef>())
+                w = getTypeWidth(decl->to<IR::Type_Typedef>(), typeMap, refMap);
+            return w;
+        }
+    }
+    LOG3("End " << type->getP4Type());
+    return type->width_bits();
+}
+
+/*
+ * Function only supports single-line type statement and if p4runtime Annotation
```

Comment:
I don't know what "single-line type statement" means. If you mean a simple `type <built-in type> <type name>` declaration, it is irrelevant. As you yourself said in the past, any level of nesting should be supported.

Reply:
My plan is to use a subsequent PR to support any level of nesting since it is complex.  The complexity is covered in Andy's document and Example 3. 

https://github.com/jafingerhut/p4-guide/blob/master/p4runtime/README-p4info-and-type.md#proposed-p4c-rules-for-generating-types-in-p4info-files

First, it would be good to get the other changes in.

Reply:
By "single-line type statement" I do mean a simple `type <built-in type> <type name>` declaration.